collector/ilm_indices.go

```diff
@@ -0,0 +1,164 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type ilmMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(timeMillis float64) float64
+	Labels []string
+}
+
+// Index Lifecycle Management information object
+type IlmIndiciesCollector struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	up                prometheus.Gauge
+	totalScrapes      prometheus.Counter
+	jsonParseFailures prometheus.Counter
+
+	ilmMetric ilmMetric
+}
+
+var (
+	defaultIlmIndicesMappingsLabels = []string{"index", "phase", "action", "step"}
+)
+
+// NewIlmIndicies defines Index Lifecycle Management Prometheus metrics
+func NewIlmIndicies(logger log.Logger, client *http.Client, url *url.URL) *IlmIndiciesCollector {
+	subsystem := "ilm_index"
+
+	return &IlmIndiciesCollector{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "up"),
+			Help: "Was the last scrape of the ElasticSearch ILM endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "total_scrapes"),
+			Help: "Current total ElasticSearch ILM scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		ilmMetric: ilmMetric{
+			Type: prometheus.GaugeValue,
+			Desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, subsystem, "status"),
+				"Status of ILM policy for index",
+				defaultIlmIndicesMappingsLabels, nil),
+			Value: func(timeMillis float64) float64 {
+				return timeMillis
+			},
+		},
+	}
+}
+
+// Describe adds metrics description
+func (i *IlmIndiciesCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- i.ilmMetric.Desc
+	ch <- i.up.Desc()
+	ch <- i.totalScrapes.Desc()
+	ch <- i.jsonParseFailures.Desc()
+}
+
+func (i *IlmIndiciesCollector) fetchAndDecodeIlm() (IlmResponse, error) {
+	var ir IlmResponse
+
+	u := *i.url
+	u.Path = path.Join(u.Path, "/_all/_ilm/explain")
+
+	res, err := i.client.Get(u.String())
+	if err != nil {
+		return ir, fmt.Errorf("failed to get index stats from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			_ = level.Warn(i.logger).Log(
+				"msg", "failed to close http.Client",
+				"err", err,
+			)
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		return ir, fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	if err := json.NewDecoder(res.Body).Decode(&ir); err != nil {
```

Comment:
I think this should use the readall/unmarshal as the other file. `json.NewDecoder().Decode()` can lead to unexpected behavior.